---
title: "Genomic_Selection"
author: "Guilherme, Mandeep & Subash Thapa"
date: "2024-11-08"
output: html_document
---

# Part v) Genomic selection

For this part, it is important to reminder that we are using a new dataset that fits better for the genomic selection analysis.

## Checking data stuctures

Before running any analysis, it is important to check the data structures, to avoid mistakes and warnings in the R code. 

```{r}
str(Data.Maize)
str(Gg)
```

In this case, the REP column is not a factor, so, we need to fix it.

```{r}
Data.Maize$Rep <- as.factor(Data.Maize$Rep)
str(Data.Maize)
```

## Checking data distribution 

```{r}
ggplot(Data.Maize, aes(x = Env, y = Yield, fill = Env)) +
  geom_boxplot() +
  labs(title = "Yield by Environment", x = "Environment", y = "Yield") +
  scale_fill_manual(values = c("grey5", "grey55", "grey95")) + 
  theme_minimal()

ggplot(Data.Maize, aes(x = Env, y = ASI, fill = Env)) +
  geom_boxplot() +
  labs(title = "ASI by Environment", x = "Environment", y = "ASI") +
  scale_fill_manual(values = c("grey5", "grey55", "grey95")) + 
  theme_minimal()

ggplot(Data.Maize, aes(x = Env, y = PH, fill = Env)) +
  geom_boxplot() +
  labs(title = "PH by Environment", x = "Environment", y = "PH") +
  scale_fill_manual(values = c("grey5", "grey55", "grey95")) + 
  theme_minimal()
```

## Correlation between enviroments

```{r}
yield_wide <- Data.Maize %>%
  select(Gid, Env, Yield) %>%  
  spread(key = Env, value = Yield)

correlation_matrix <- yield_wide %>%
  select(-Gid) %>%  
  cor(use = "complete.obs") 

print(correlation_matrix)
```

## Correlation between traits

```{r}
compute_correlations <- function(env_data) {
  cor(env_data %>% select(Yield, ASI, PH), use = "complete.obs")
}

correlations_by_env <- Data.Maize %>%
  group_by(Env) %>%
  summarise(correlation_matrix = list(compute_correlations(cur_data())))


correlations_by_env %>%
  filter(Env == "EBU") %>%
  pull(correlation_matrix) %>%
  .[[1]]  

correlations_by_env %>%
  filter(Env == "KAK") %>%
  pull(correlation_matrix) %>%
  .[[1]]  

correlations_by_env %>%
  filter(Env == "KTI") %>%
  pull(correlation_matrix) %>%
  .[[1]]  
```

## Fixing data strcture (again) 

```{r}
Data.Maize <- (Data.Maize[order(Data.Maize$Env,Data.Maize$Gid),])
rownames(Data.Maize)=1:nrow(Data.Maize)
head(Data.Maize)
```

## Design of matrices

```{r}
LG <- cholesky(Gg)
ZG <- model.matrix(~0 + as.factor(Data.Maize$Gid))
Z.G <- ZG %*% LG
Z.E <- model.matrix(~0 + as.factor(Data.Maize$Env))
ZEG <- model.matrix(~0 + as.factor(Data.Maize$Gid):as.factor(Data.Maize$Env))
G2 <- kronecker(diag(length(unique(Data.Maize$Env))),data.matrix(Gg))
LG2 <- cholesky(G2)
Z.EG <- ZEG %*% LG2
Y <- as.matrix(Data.Maize[, -c(1, 2, 3)])
```

## Fitting the model 

```{r}
fm <- BMTME(Y = Y, X = Z.E, Z1 = Z.G, Z2 = Z.EG,
#nIter =15000, burnIn =10000, thin = 2,bs = 50)
nIter =150, burnIn =100, thin = 2,bs = 50)
```

## Extracting covariances

```{r}
COV_TraitGenetic <- fm$varTrait
COV_TraitGenetic
```

## Correlation matrix

```{r}
COR_TraitGenetic <- cov2cor(COV_TraitGenetic)
COR_TraitGenetic
```

## Correlation matrix between environments

```{r}
COV_EnvGenetic <- fm$varEnv
COV_EnvGenetic
```

## Correlation matrix between traits

```{r}
COV_ResGenetic <- fm$vare
COV_ResGenetic
```

## Predictions

```{r}
plot(fm, trait="Yield")
```

## Cross validation

```{r}
pheno <- data.frame(GID = Data.Maize[, 1], Env =
Data.Maize[, 2], Response = Data.Maize[, 4])
CrossV <- CV.KFold(pheno, DataSetID = "GID", K = 5,
set_seed = 123)
pm <- BMTME(Y = Y, X = Z.E, Z1 = Z.G, Z2 = Z.EG,
nIter = 250, burnIn = 50, thin = 2,bs = 50, testingSet = CrossV)
```

## Final results

```{r}
summary(pm)
```

# References

*Crossa, J., Y. Beyene, S. Kassa, P. Pérez-Rodríguez, J. M. Hickey, et al., 2013 Genomic prediction in maize breeding populations with genotyping-by-sequencing. G3: Genes|Genomes|Genetics (Bethesda) 3, 1903–1926. https://doi.org/10.1534/g3.113.008227*

*Dias, K.O.D.G., Gezan, S.A., Guimarães, C.T. et al. Improving accuracies of genomic predictions for drought tolerance in maize by joint modeling of additive and dominance effects in multi-environment trials. Heredity 121, 24–37 (2018). https://doi.org/10.1038/s41437-018-0053-6*

*Montesinos-López, O. A., A. Montesinos-López, J. Crossa, F. Toledo,O. Pérez-Hernández et al., 2016 A Genomic Bayesian Multi-trait and Multi-environment model. G3: Genes|Genomes|Genetics (Bethesda),6:2725–2744. https://doi.org/10.1534/g3.116.032359*

*Montesinos-López, O. A., A. Montesinos-López, J. Crossa, J. C. Montesinos-López, F. J. Luna-Vázquez et al., 2017 A Variational Bayes Genomic-Enabled Prediction Method with Genotype · Environment Interaction. G3: Genes, Genomes. Genetics 7: 1833–1853*

*Osval A Montesinos-López, Abelardo Montesinos-López, Francisco Javier Luna-Vázquez, Fernando H Toledo, Paulino Pérez-Rodríguez, Morten Lillemo, José Crossa, An R Package for Bayesian Analysis of Multi-environment and Multi-trait Multi-environment Data for Genome-Based Prediction, G3 Genes|Genomes|Genetics, Volume 9, Issue 5, 1 May 2019, Pages 1355–1369, https://doi.org/10.1534/g3.119.400126*

*Olivoto T, Lúcio AD. metan: An R package for multi-environment trial analysis. Methods Ecol Evol. 2020; 11: 783–789. https://doi.org/10.1111/2041-210X.13384*
