---
title: "Genomic Selection Analysis"
author: "Guilherme Oliveira, Mandeep Singh, Subash Thapa"
date: "2024-11-05"
output:
  html_document: default
  pdf_document: default
---

The following document was developed by Guilherme Oliveira, Mandeep Singh, Subash Thapa, students at South Dakota State University, as a requisite for the course PS 746 - Quantitative Genetics. 

The main objective of this project was to conduct Multi Trial Analysis, covering i) GxE analysis (+ GGE analysis) ii) AMMI analysis, iii) GGI analysis, iv) FW analysis and v) genomic selection approach which explores in some way Genotype x Environment interaction (G x E).

For the parts i, ii, iii, iv, we are exploring the open dataset made available by Dias et al. (2018) which contains phenotypic data of five drought tolerance traits, measured in in 308 hybrids along eight environments contrasting for water availability. For practical purposes, we are using a subset of 202 hybrids. The traits analyzed were grain yield (GY), ears per plot (EPP), female and male flowering times (FFT and MFT), and anthesis-silking interval (ASI). The source code used is from the metan (Olivoto and Lucio, 2020) and statgenGxE reference manuals. 

For the part v, we are exploring the open dataset made available by Crossa et al. 2013 and Montesinos-Lopez et al. 2016, 2017 which contains the data of 309 double-haploid maize lines conducted in 3 environments and each environments having 3 reps of each line. The traits analyzed were grain yield (yield), anthesis-silking interval (ASI) and plant height (PH). The source code used is from Montesinos-Lopez et al. 2019 using the BMTME R package, where we did just small modifications, however, bringing more interpretations and extra analyses.

## Directory

```{r}
setwd("C:/Users/Guilherme.Oliveira/Documents/PS 756 - Final Project")
getwd()
```

#Genomic selection

For this part, it is important to reminder that we are using a new dataset that fits better for the genomic selection analysis.

## Checking data stuctures

Before running any analysis, it is important to check the data structures, to avoid mistakes and warnings in the R code. 

```{r}
str(Data.Maize)
str(Gg)
```

In this case, the REP column is not a factor, so, we need to fix it.

```{r}
Data.Maize$Rep <- as.factor(Data.Maize$Rep)
str(Data.Maize)
```

## Checking data distribution 

In this piece of code as we did for the other dataset, we are checking the distribution and outliers.

```{r}
ggplot(Data.Maize, aes(x = Env, y = Yield, fill = Env)) +
  geom_boxplot() +
  labs(title = "Yield by Environment", x = "Environment", y = "Yield") +
  scale_fill_manual(values = c("grey5", "grey55", "grey95")) + 
  theme_minimal()

ggplot(Data.Maize, aes(x = Env, y = ASI, fill = Env)) +
  geom_boxplot() +
  labs(title = "ASI by Environment", x = "Environment", y = "ASI") +
  scale_fill_manual(values = c("grey5", "grey55", "grey95")) + 
  theme_minimal()

ggplot(Data.Maize, aes(x = Env, y = PH, fill = Env)) +
  geom_boxplot() +
  labs(title = "PH by Environment", x = "Environment", y = "PH") +
  scale_fill_manual(values = c("grey5", "grey55", "grey95")) + 
  theme_minimal()
```

We observe some outliers which can affect our predictions, so, for this case, we are removing them from the dataset. To complete this task, we are using the previous code showed in the beginning of this project.

```{r}
#boxplot.Yield <- boxplot(Data.Maize$Yield)
#boxplot.ASI <- boxplot(Data.Maize$ASI)
#boxplot.PH <- boxplot(Data.Maize$PH)

#outliers <- boxplot.Yield$out; outliers
#Data.Maize <- Data.Maize[-which(Data.Maize$Yield%in%outliers),]
#shapiro.test(Data.Maize$Yield)

#outliers2 <- boxplot.ASI$out; outliers2
#Data.Maize <- Data.Maize[-which(Data.Maize$ASI%in%outliers2),]
#shapiro.test(Data.Maize$ASI)

#outliers3 <- boxplot.PH$out; outliers3
#Data.Maize <- Data.Maize[-which(Data.Maize$PH%in%outliers3),]
#shapiro.test(Data.Maize$PH)
```

## Correlation between enviroments

```{r}
yield_wide <- Data.Maize %>%
  select(Gid, Env, Yield) %>%  
  spread(key = Env, value = Yield)

correlation_matrix <- yield_wide %>%
  select(-Gid) %>%  
  cor(use = "complete.obs") 

print(correlation_matrix)
```
The correlations between the environments are positive, however, it is not a very strong correlations, meaning that using all the three environments in this analysis can be a good idea.

## Correlation between traits

```{r}
compute_correlations <- function(env_data) {
  cor(env_data %>% select(Yield, ASI, PH), use = "complete.obs")
}

correlations_by_env <- Data.Maize %>%
  group_by(Env) %>%
  summarise(correlation_matrix = list(compute_correlations(cur_data())))


correlations_by_env %>%
  filter(Env == "EBU") %>%
  pull(correlation_matrix) %>%
  .[[1]]  

correlations_by_env %>%
  filter(Env == "KAK") %>%
  pull(correlation_matrix) %>%
  .[[1]]  

correlations_by_env %>%
  filter(Env == "KTI") %>%
  pull(correlation_matrix) %>%
  .[[1]]  
```
Interesting we have some positive and negative correlation among the traits. Yield and PH are positive correlated, while both traits are negative correlated to ASI. 

## Fixing data strcture (again) 

```{r}
Data.Maize <- (Data.Maize[order(Data.Maize$Env,Data.Maize$Gid),])
rownames(Data.Maize)=1:nrow(Data.Maize)
head(Data.Maize)
```

## Design of matrices

Here we have one of the most important part of this genomic selection. In this step we are developing matrices for the line effects, the environment and the genotype x environment.

So, first of all, lets check our genotypic information data.

```{r}
str(Gg)
head(Gg, n = c(5,5))
```
So, in this datatset, what we got and we are showing here is a G matrix (kinship matrix). Therefore, it is not the molecular markers code. There are different ways to get the G matrix, however, we are not spending much time and how to get this specific matrix, because, we are going more in detail on the other matrices for our genomic selection model.

So, the first step is to conduct a Cholesky factorization. It is a technique used in linear algebra, basically, to make easier next algebraic operations. Note, that our matrix has the same dimensions, however, we have different values.

```{r}
LG <- cholesky(Gg)
str(LG)
head(LG, n = c(5,5))
```
Next, we are just creating a matrix with the name of our hybrids, where we are coding according the row number, the hybrids order. Basically to associate with our previous matrix.

```{r}
ZG <- model.matrix(~0 + as.factor(Data.Maize$Gid))
str(ZG)
head(ZG, n = c(5,5))
```
So, now we have two matrices: one is our G matrix and the other one is matrix containing the genotype order and we are multiplying them. The reasons behind is that the Z.G is basically the LG x 3. We just duplicate the genotypic information by 3, because we are using this matrix as a base matrix to account for environment effects (we have three environments).

```{r}
Z.G <- ZG %*% LG
str(Z.G)
head(Z.G, n = c(5,5))
```
The purpose of this matrix is very similar to the Z.G. But here is to code the hybrids observation to a respective environment.

```{r}
Z.E <- model.matrix(~0 + as.factor(Data.Maize$Env))
str(Z.E)
head(Z.E, n = c(5,5))
```
This matrix is combining basically all the hybrids and environments.

```{r}
ZEG <- model.matrix(~0 + as.factor(Data.Maize$Gid):as.factor(Data.Maize$Env))
str(ZEG)
head(ZEG, n = c(5,5))
```
In summary, now we are creating a multi-environment  genomic relationship 

```{r}
G2 <- kronecker(diag(length(unique(Data.Maize$Env))),data.matrix(Gg))
LG2 <- cholesky(G2)
str(LG2)
head(LG2, n = c(5,5))
```
We are transforming the genotype-environment interaction matrix according to the multi-environment genomic relationships.

```{r}
Z.EG <- ZEG %*% LG2
str(Z.EG)
head(Z.EG, n = c(5,5))
```

```{r}
Y <- as.matrix(Data.Maize[, -c(1, 2, 3)])
str(Y)
head(Y, n = c(5,5))
```

## Fitting the model 

Model interpretation:
Y = Phenotypic trait data for each genotype-environment combination.
X  (Z.E) = Environment effects in the model
Z1 (Z.G.) = Genetic effects, combining genotype information with genetic relationships
Z2 (Z.EG) = Genotype-by-environment (GxE) interaction effects
nIter = Total number of Markov Chain Monte Carlo iterations for sampling posterior distribution. High number is required to ensure model convergence
burnIn = Number of initial iterations that will be discarded (chain stabilization and avoiding bias)
Thin = Determines that every second iteration will be kept (others discarded). Thinning reduces autocorrelation in the samples.
bs = Number of posterior samples to save.


```{r}
LG <- cholesky(Gg)
ZG <- model.matrix(~0 + as.factor(Data.Maize$Gid))
Z.G <- ZG %*% LG
Z.E <- model.matrix(~0 + as.factor(Data.Maize$Env))
ZEG <- model.matrix(~0 + as.factor(Data.Maize$Gid):as.factor(Data.Maize$Env))
G2 <- kronecker(diag(length(unique(Data.Maize$Env))),data.matrix(Gg))
LG2 <- cholesky(G2)
Z.EG <- ZEG %*% LG2
Y <- as.matrix(Data.Maize[, -c(1, 2, 3)])
fm <- BMTME(Y = Y, X = Z.E, Z1 = Z.G, Z2 = Z.EG,
#nIter =15000, burnIn =10000, thin = 2,bs = 50)
nIter =150, burnIn =100, thin = 2,bs = 50)
```

## Extracting covariances

```{r}
COV_TraitGenetic <- fm$varTrait
COV_TraitGenetic
```

## Covariance matrix between traits

```{r}
COR_TraitGenetic <- cov2cor(COV_TraitGenetic)
COR_TraitGenetic
```

## Covariance matrix between environments

```{r}
COV_EnvGenetic <- fm$varEnv
COV_EnvGenetic
```

## Residual covariance matrix between traits

```{r}
COV_ResGenetic <- fm$vare
COV_ResGenetic
```

## Predictions

```{r}
plot(fm, trait="Yield")
```

## Cross validation

```{r}
pheno <- data.frame(GID = Data.Maize[, 1], Env =
Data.Maize[, 2], Response = Data.Maize[, 4])
CrossV <- CV.KFold(pheno, DataSetID = "GID", K = 5,
set_seed = 123)
pm <- BMTME(Y = Y, X = Z.E, Z1 = Z.G, Z2 = Z.EG,
nIter = 250, burnIn = 50, thin = 2,bs = 50, testingSet = CrossV)
```

## Final results

```{r}
summary(pm)
boxplot(pm, select = "MAAPE", las = 2)
```

# References

*Crossa, J., Y. Beyene, S. Kassa, P. Pérez-Rodríguez, J. M. Hickey, et al., 2013 Genomic prediction in maize breeding populations with genotyping-by-sequencing. G3: Genes|Genomes|Genetics (Bethesda) 3, 1903–1926. https://doi.org/10.1534/g3.113.008227*

*Dias, K.O.D.G., Gezan, S.A., Guimarães, C.T. et al. Improving accuracies of genomic predictions for drought tolerance in maize by joint modeling of additive and dominance effects in multi-environment trials. Heredity 121, 24–37 (2018). https://doi.org/10.1038/s41437-018-0053-6*

*Montesinos-López, O. A., A. Montesinos-López, J. Crossa, F. Toledo,O. Pérez-Hernández et al., 2016 A Genomic Bayesian Multi-trait and Multi-environment model. G3: Genes|Genomes|Genetics (Bethesda),6:2725–2744. https://doi.org/10.1534/g3.116.032359*

*Montesinos-López, O. A., A. Montesinos-López, J. Crossa, J. C. Montesinos-López, F. J. Luna-Vázquez et al., 2017 A Variational Bayes Genomic-Enabled Prediction Method with Genotype · Environment Interaction. G3: Genes, Genomes. Genetics 7: 1833–1853*

*Osval A Montesinos-López, Abelardo Montesinos-López, Francisco Javier Luna-Vázquez, Fernando H Toledo, Paulino Pérez-Rodríguez, Morten Lillemo, José Crossa, An R Package for Bayesian Analysis of Multi-environment and Multi-trait Multi-environment Data for Genome-Based Prediction, G3 Genes|Genomes|Genetics, Volume 9, Issue 5, 1 May 2019, Pages 1355–1369, https://doi.org/10.1534/g3.119.400126*

*Olivoto T, Lúcio AD. metan: An R package for multi-environment trial analysis. Methods Ecol Evol. 2020; 11: 783–789. https://doi.org/10.1111/2041-210X.13384*